### PR TITLE
Implement `endBefore` for RTDB queries

### PR DESCRIPTION
**Pull request by IanWyszynski**
go/rtdb-exclusive-query-filters

### CODE REVIEW

**Comment by IanWyszynski** [comment]
@schmidt-sebastian Ready for review.

**Comment by IanWyszynski** [comment]
@schmidt-sebastian Do you have any thoughts on whether we should merge this PR into https://github.com/firebase/firebase-android-sdk/pull/2277 and then merge the two to master?

File: firebase-database/src/test/java/com/google/firebase/database/PushIdGeneratorTest.java [hunk 1]
```diff
@@ -56,4 +57,21 @@ public void testSuccessorBasic() {
     assertEquals(
         PushIdGenerator.successor("abc" + MIN_PUSH_CHAR), "abc" + MIN_PUSH_CHAR + MIN_PUSH_CHAR);
   }
+
+  @Test
+  public void testPredecessorSpecialValue() {
+    assertEquals(
+        PushIdGenerator.predecessor(String.valueOf(MIN_PUSH_CHAR)),
+        String.valueOf(Integer.MAX_VALUE));
```
Comment (schmidt-sebastian) [inline firebase-database/src/test/java/com/google/firebase/database/PushIdGeneratorTest.java:65]: Same comment as in other PR: These arguments should be flipped.

File: firebase-database/src/main/java/com/google/firebase/database/Query.java [hunk 2]
```diff
@@ -478,6 +478,96 @@ private Query startAt(Node node, String key) {
     return new Query(repo, path, newParams, orderByCalled);
   }
 
+  /**
+   * Create a query constrained to only return child nodes with a value greater than the given
+   * value, using the given orderBy directive or priority as default.
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/Query.java:483]: I just realized that all these comments say "value greater than", when it should be "value less than".

File: firebase-database/src/main/java/com/google/firebase/database/Query.java [hunk 3]
```diff
@@ -478,6 +478,96 @@ private Query startAt(Node node, String key) {
...
+  public Query endBefore(boolean value) {
+    return endAt(value, ChildKey.getMinName().asString());
+  }
+
+  /**
+   * Create a query constrained to only return child nodes with a value greater than the given
+   * value, using the given orderBy directive or priority as default, and additionally only child
+   * nodes with a key greater than or equal to the given key.
+   *
+   * @param value The priority to start at
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/Query.java:525]: s/value/priority

File: firebase-database/src/main/java/com/google/firebase/database/Query.java [hunk 4]
```diff
@@ -478,6 +478,96 @@ private Query startAt(Node node, String key) {
...
+        value != null ? new StringNode(value, PriorityUtilities.NullPriority()) : EmptyNode.Empty();
+    return endBefore(node, key);
+  }
+
+  /**
+   * Create a query constrained to only return child nodes with a value greater than the given
+   * value, using the given orderBy directive or priority as default, and additionally only child
+   * nodes with a key greater than or equal to the given key.
+   *
+   * @param value The priority to start at
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/Query.java:542]: s/value/priority

File: firebase-database/src/main/java/com/google/firebase/database/Query.java [hunk 5]
```diff
@@ -478,6 +478,96 @@ private Query startAt(Node node, String key) {
...
+  public Query endBefore(double value, @Nullable String key) {
+    return endBefore(new DoubleNode(value, PriorityUtilities.NullPriority()), key);
+  }
+
+  /**
+   * Create a query constrained to only return child nodes with a value greater than the given
+   * value, using the given orderBy directive or priority as default, and additionally only child
+   * nodes with a key greater than or equal to the given key.
+   *
+   * @param value The priority to start at
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/Query.java:557]: s/value/priority

**Review by schmidt-sebastian — COMMENTED** [review]
I don't mind how you merge these PRs. It is likely a bit easier to merge this into `startAfter` and then merge together, especially given the CI flakes.

**Review by schmidt-sebastian — APPROVED** [review]

**Pull request merged**